cyclic_boosting/regression.py

```diff
@@ -180,4 +183,246 @@ def calc_parameters(self, feature, y, pred, prefit_data):
+    structure allows arbitrary/empirical target ranges/distributions, the
+    multiplicative model of this mode requires non-negative target values.
+
+    A quantile loss, according to the desired quantile to be predicted, is
+    minimized in each bin of each feature. While binning, feature cycles,
+    smoothing, and iterations work in the same way as usual in Cyclic Boosting,
+    the minimization itself is performed via ``scipy.optimize.minimize``
+    (instead of an analytical solution like, e.g., in ``CBPoissonRegressor``,
+    ``CBNBinomRegressor``, or ``CBLocationRegressor``).
+
+    Parameters
+    ----------
+    quantile : float
+        quantile to be estimated
+    See :class:`cyclic_boosting.base` for all other parameters.
+    """
+
+    def __init__(
+        self,
+        feature_groups=None,
+        feature_properties=None,
+        weight_column=None,
+        prior_prediction_column=None,
+        minimal_loss_change=1e-10,
+        minimal_factor_change=1e-10,
+        maximal_iterations=10,
+        observers=None,
+        smoother_choice=None,
+        output_column=None,
+        learn_rate=None,
+        quantile=0.5,
+        aggregate=True,
+    ):
+        CyclicBoostingBase.__init__(
+            self,
+            feature_groups=feature_groups,
+            feature_properties=feature_properties,
+            weight_column=weight_column,
+            prior_prediction_column=prior_prediction_column,
+            minimal_loss_change=minimal_loss_change,
+            minimal_factor_change=minimal_factor_change,
+            maximal_iterations=maximal_iterations,
+            observers=observers,
+            smoother_choice=smoother_choice,
+            output_column=output_column,
+            learn_rate=learn_rate,
+            aggregate=aggregate,
+        )
+
+        self.quantile = quantile
+
+    def precalc_parameters(self, feature, y, pred):
+        pass
+
+    def loss(self, prediction, y, weights):
+        """
+        Calculation of the in-sample quantile loss, or to be exact costs,
+        (potentially including sample weights) after full feature cycles, i.e.,
+        iterations, to be used as stopping criteria.
+
+        Parameters
+        ----------
+        prediction : np.ndarray
+            (in-sample) predictions for desired quantile, containing data with `float` type
+        y : np.ndarray
+            target variable, containing data with `float` type (potentially discrete)
+        weights : np.ndarray
+            optional (otherwise set to 1) sample weights, containing data with `float` type
+
+        Returns
+        -------
+        float
+            calcualted quantile costs
+        """
+        if not len(y) > 0:
+            raise ValueError("Loss cannot be computed on empty data")
+        else:
+            sum_weighted_error = np.nansum(
+                (
+                    (y < prediction) * (1 - self.quantile) * (prediction - y)
```

Comment:
Because the expression for the quantile_loss is used here and in https://github.com/Blue-Yonder-OSS/cyclic-boosting/pull/35/files#diff-b9b12e19544ab33f1d7e4d74b8b8b271db2481edf58a4f64727b1f497af111fcR339 , one could write a helper function with
```
def quantile_loss(y,y_pred,quantile):
      return (y < y_pred) * (1 - quantile) * (y_pred - y) + (y >= y_pred) * quantile * (y - y_pred)
```

Reply:
In principle, yes. But here it is the loss after full feature iterations and in the actual optimization step it is for each feature bin containing the specific parameter to be estimated.
But let me see if we can improve this in the upcoming generalizing pull request.